### PR TITLE
feat(SortedList): abstract SortedList

### PR DESCRIPTION
**Pull request by Khaazz**
# PULL REQUEST

## **Overview**

This PR abstract (and remove) `TimeoutQueue` introduced in #65 in its own structure called `SortedList`.
This makes it a more general structure that can be used in more situation. This doesn't change the implementation of Collectors however.

## **Status**

- [X] Typings have been updated or don't need to be.
- [X] This PR have been tested and is ready to be merged.

## **Semantic versioning classification**

- [ ] This PR introduces *BREAKING* changes.
- [X] This PR adds new features, improve the code and implies minimal changes.
- [ ] This PR fixes a bug and references the relevant issue or documentation.
- [X] This PR improve performance or code refactor without API changes.
- [ ] This PR **only** includes non-code changes (documentation, CI, tools...).

### CODE REVIEW

File: src/Utility/External/SortedList.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ *
+ * @class SortedList
+ * @prop {Boolean} acending
+ * @prop {(toInsert, baseElement) => Boolean)} comparator - Wether toInsert is bigger than baseElement
+ * @prop {Array<T>} list
+ */
+class SortedList {
+    constructor(comparator = null, ascending = true) {
+        console.log(comparator);
+        this.comporator = comparator || ( (toInsert, baseElement) => toInsert >= baseElement);
```
Comment (bsian03) [inline src/Utility/External/SortedList.js]:
```suggestion
        this.comparator = comparator || ( (toInsert, baseElement) => toInsert >= baseElement);
```
khaaz is having a moment lol

File: src/Utility/External/SortedList.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
+/**
+ * SortedList Object. Sorted Array of element (ascending or descending order)
+ *
+ * @author KhaaZ
+ *
+ * @template T
+ *
+ * @class SortedList
+ * @prop {Boolean} acending
+ * @prop {(toInsert, baseElement) => Boolean)} comparator - Wether toInsert is bigger than baseElement
```
Comment (bsian03) [inline src/Utility/External/SortedList.js]: toInsert and baseElement types

Comment (bsian03) [reply src/Utility/External/SortedList.js]:
```suggestion
 * @prop {(toInsert, baseElement) => Boolean)} comparator - Whether toInsert is bigger than baseElement
```

File: src/Utility/External/SortedList.js (hunk 1 above)
Comment (Khaazz) [reply src/Utility/External/SortedList.js]: Apparently I also used `comporotor` in  add lol 💯

File: src/Utility/External/SortedList.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    /**
+     * Adds an element in the queue, sorting by ascending element
+     *
+     * @param {T} element
+     * @param {(toInsert, baseElement) => Boolean} comparator
+     * @memberof SortedList
+     */
+    add(element, comparator = null) {
+        const comp = comparator || this.comporator;
```
Comment (Khaazz) [inline src/Utility/External/SortedList.js]:
```suggestion
        const comp = comparator || this.comparator;
```

File: src/Utility/External/SortedList.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
+/**
+ * SortedList Object. Sorted Array of element (ascending or descending order)
+ *
+ * @author KhaaZ
+ *
+ * @template T
+ *
+ * @class SortedList
+ * @prop {Boolean} acending
+ * @prop {(toInsert, baseElement) => Boolean)} comparator - Whether toInsert is bigger than baseElement
```
Comment (bsian03) [inline src/Utility/External/SortedList.js]: Still need `toInsert` and `baseElement` types here

Comment (Khaazz) [reply src/Utility/External/SortedList.js]: which is `T` right?

Comment (bsian03) [reply src/Utility/External/SortedList.js]: I presume so

File: src/Utility/External/SortedList.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        this.comparator = comparator || ( (toInsert, baseElement) => toInsert >= baseElement);
+        this.ascending = ascending;
+        this.list = [];
+    }
+
+    /**
+     * Size of the list
+     *
+     * @readonly
+     * @typedef {Number}
```
Comment (bsian03) [inline src/Utility/External/SortedList.js]:
```suggestion
     * @type {Number}
```

File: src/Utility/External/SortedList.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+    }
+
+    /**
+     * Returns the first element of the List without removing it
+     *
+     * @returns {T}
+     * @memberof SortedList
+     */
+    last() {
+        return this.list[this.list.length];
```
Comment (bsian03) [inline src/Utility/External/SortedList.js]:
```suggestion
        return this.list[this.list.length - 1];
```

File: src/Utility/External/SortedList.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * @returns {T}
+     * @memberof SortedList
+     */
+    first() {
+        return this.list[0];
+    }
+
+    /**
+     * Returns the first element of the List without removing it
```
Comment (bsian03) [inline src/Utility/External/SortedList.js]:
```suggestion
     * Returns the last element of the list without removing it
```
Also lowercase every `List` please, consistency

**Pull request merged**